src/perf_mainloop.cc

```diff
@@ -85,47 +120,69 @@ static void pollfd_setup(const PEventHdr *pevent_hdr, struct pollfd *pfd,
   }
 }
 
-static void worker(DDProfContext *ctx, const WorkerAttr *attr,
-                   MLWorkerFlags *flags) {
-  // Until a restartable terminal condition is met, the worker will set its
-  // disposition so that profiling is halted upon its termination
-  flags->can_run = false;
-  flags->errors = true;
+static DDRes signalfd_setup(pollfd *pfd) {
+  sigset_t mask;
+
+  sigemptyset(&mask);
+  sigaddset(&mask, SIGINT);
+  sigaddset(&mask, SIGTERM);
+  
+  // no need to block signal, since we inherited sigprocmask from parent
+  int sfd = signalfd(-1, &mask, 0);
+  DDRES_CHECK_ERRNO(sfd, DD_WHAT_WORKERLOOP_INIT, "Could not set signalfd");
+
+  pfd->fd = sfd;
+  pfd->events = POLLIN;
+  return {};
+}
+
+static DDRes worker_loop(DDProfContext *ctx, const WorkerAttr *attr,
+                         bool *restart_worker) {
 
   // Setup poll() to watch perf_event file descriptors
   int pe_len = ctx->worker_ctx.pevent_hdr.size;
-  struct pollfd pfd[MAX_NB_WATCHERS];
+  // one extra slot in pfd to accomodate for signal fd
+  struct pollfd pfd[MAX_NB_WATCHERS + 1];
   int pfd_len = 0;
   pollfd_setup(&ctx->worker_ctx.pevent_hdr, pfd, &pfd_len);
 
+  DDRES_CHECK_FWD(signalfd_setup(&pfd[pfd_len]));
+  int signal_pos = pfd_len++;
+
   // Perform user-provided initialization
-  DDRES_CHECK_OR_SHUTDOWN(attr->init_fun(ctx), attr->finish_fun(ctx));
+  defer { attr->finish_fun(ctx); };
+  DDRES_CHECK_FWD(attr->init_fun(ctx));
 
   // Worker poll loop
-  while (1) {
+  while (true) {
     uint64_t processed_samples = 0;
     int n = poll(pfd, pfd_len, PSAMPLE_DEFAULT_WAKEUP);
 
     // If there was an issue, return and let the caller check errno
     if (-1 == n && errno == EINTR) {
       continue;
-    } else if (-1 == n) {
-      DDRES_CHECK_OR_SHUTDOWN(ddres_error(DD_WHAT_POLLERROR),
-                              attr->finish_fun(ctx));
+    }
+    DDRES_CHECK_ERRNO(n, DD_WHAT_POLLERROR, "poll failed");
+
+    if (pfd[signal_pos].revents & POLLIN) {
+      struct signalfd_siginfo siginfo;
+      if (read(pfd[signal_pos].fd, &siginfo, sizeof(siginfo)) !=
```

Comment:
So, the `signalfd()` was set on a mask comprised of SIGINT and SIGTERM.  Since the only thing we can do at this point is possibly fail to read the value (and exit anyway, except with an error unrelated to the original termination of the process), is it possible to just rely on the presence of `pfd[signal_pos].revents & POLLIN` as the canonical indicator that a shutdown signal was received?

Reply:
Yup,that's right!